pydantic/_internal/_generate_schema.py

```diff
@@ -911,17 +919,33 @@ def _std_types_schema(self, obj: Any) -> core_schema.CoreSchema | None:
             except KeyError:
                 continue
             return encoder(self, obj)
-        if dataclasses.is_dataclass(obj):
-            return self._dataclass_schema(obj)  # type: ignore
         return None
 
-    def _dataclass_schema(self, dataclass: type[StandardDataclass]) -> core_schema.CoreSchema:
+    def _dataclass_schema(
+        self, dataclass: type[StandardDataclass], origin: type[StandardDataclass] | None
+    ) -> core_schema.CoreSchema:
         """
         Generate schema for a dataclass.
         """
+
+        if origin is not None:
+            # In this case, we know that dataclass is a _GenericAlias, and origin is the generic type
+            # So it is safe to access dataclass.__args__ and origin.__parameters__
+            args: tuple[Any, ...] = dataclass.__args__  # type: ignore
+            parameters: tuple[_typing_extra.TypeVarType, ...] = origin.__parameters__  # type: ignore
+            typevars_map = dict(zip(parameters, args))
+            dataclass = origin
+        else:
+            typevars_map = None
+
         # FIXME we need a way to make sure kw_only info is propagated through to fields
         fields, _ = collect_fields(
-            dataclass, dataclass.__bases__, self.types_namespace, dc_kw_only=True, is_dataclass=True
+            dataclass,
+            dataclass.__bases__,
+            self.types_namespace,
+            dc_kw_only=True,
+            is_dataclass=True,
+            typevars_map=typevars_map,
         )
```

Comment:
I agree with @samuelcolvin that if we have an `is_dataclass: bool` param we might just want two functions.

Reply:
Yeah, he and I talked about it earlier, I'm redoing things now to move in that direction.